src/lib/server/links/types.ts

```diff
@@ -0,0 +1,10 @@
+export interface ParsedLink {
+  relativePathname: string;
+  query: string;
+  hash: string;
+}
+
+export interface Link<ID extends string> extends ParsedLink {
+  id: ID;
```

Comment:
how come and ID that is a string and not just string

Reply:
The type denoted `ID` is used to define `Link['id']`. We expect to use a simple JavaScript object to store the data for all of our official shortlinks. For example, let's say we have the following shortlinks represented in our shortlinks data.

<https://github.com/EthanThatOneKid/acmcsuf.com/blob/697014c6ae2d4d6c4a095ab8b9f9d95e9e515adb/src/lib/server/links/utils.test.ts#L4-L7>

We want the data we are working with to be as accurate as possible, so I used `ID extends string` to guarantee that only string-like types can be used to represent `ID`. In practice, the `ID` gets automatically inferred from the keys of the the official shortlinks collection, `LINKS`.

For instance, in the following line of code, `ID` is inferred to be represented by the keys of the test collection of shortlinks, `TEST_LINKS`, where `keyof TEST_LINKS` evaluates to `'example' | 'example/abc'` and is automatically used to represent `ID`.
<https://github.com/EthanThatOneKid/acmcsuf.com/blob/697014c6ae2d4d6c4a095ab8b9f9d95e9e515adb/src/lib/server/links/utils.test.ts#L212>

Reply:
> how come and ID that is a string and not just string

The type denoted `ID` is used to define `Link['id']`. We expect to use a simple JavaScript object to store the data for all of our official shortlinks. For example, let's say we have the following shortlinks represented in our shortlinks data.

<https://github.com/EthanThatOneKid/acmcsuf.com/blob/697014c6ae2d4d6c4a095ab8b9f9d95e9e515adb/src/lib/server/links/utils.test.ts#L4-L7>

We want the data we are working with to be as accurate as possible, so I used `ID extends string` to guarantee that only string-like types can be used to represent `ID`. In practice, the `ID` gets automatically inferred from the keys of the the official shortlinks collection, `LINKS`.

For instance, in the following line of code, `ID` is inferred to be represented by the keys of the test collection of shortlinks, `TEST_LINKS`, where `keyof TEST_LINKS` evaluates to `'example' | 'example/abc'` and is automatically used to represent `ID`.
<https://github.com/EthanThatOneKid/acmcsuf.com/blob/697014c6ae2d4d6c4a095ab8b9f9d95e9e515adb/src/lib/server/links/utils.test.ts#L212>

What this example does not go on to show is the usefulness of inference in TypeScript considering how the return type of `parseLinkId(data, TEST_LINKS)` will be of the appropriate type as inferred by the type of the keys of `TEST_LINKS`.

Reply:
To clarify, Jason is asking why `Link` is defined like so:
<https://github.com/EthanThatOneKid/acmcsuf.com/blob/697014c6ae2d4d6c4a095ab8b9f9d95e9e515adb/src/lib/server/links/types.ts#L7-L10>

...When it could be defined like this:
```ts
export interface Link extends ParsedLink {
  id: string;
  destination: string;
}
```